### DISCUSSION TITLE
validate packet integrety

### DISCUSSION THREAD

**Discussion by geve09** [johanmeijer/grott#364] [Ideas]
In verbose or debug mode validate the package length and the data checksum. 
The package load length in in header[8:12]. The length start after the length field, upto the checksum.
The checksum (crc16 modbus 0xa001) is calculated over the whole data packet and added to the end of the packet.
"000c00060003510447a419"
Package number = 000c
Package version type = 0006
Length = 0003
Data content = 510447
- Package sub type = 5104
- Decoded sub content 47
Checksum = a419 (calculated over 000c00060003510447)

**Comment by johanmeijer** [comment]
What is your point? Grott and Grottserver are already validating crc.

**Comment by geve09** [comment]
I'm running grottest.py
With
conf.verbose = True
xdata ="000c0006024151040c252446223e4375283b7761747447726f7761747447726f7761747447723f2e2d44370f4a5f4f56747447726f7761747447726f7761747447726f7761747447726f756acc7873726e7761db3348aa6f1b6174db0075f77761747447726f7761747447726f7761747447726f776174db00726f7e217474ebb37cfe685e747a726f40877d36474c6f7758127dcf725177614f6c5769608f715a7447726f776174744772c030616f93b9726f77367474558c6f7773eb74477238776166eb47726f7761747447726f7761747447726f7761747447726f7761747447726f7761747410726f65fe747447726f776174744772387761747467ca6fdc61aa74b6726f6fb8766a453e6d73617475aa7ea77b713a5447256f77a2247447726f7761747447726f7761747447726f77617474474e6f7761747447726f77617474e8356f77612d78727ede7761747447726f7761747447726f7761747447726f776174744761437761742d47727c5b61747447726f7761747447726f7761747447726f7761747447726f7761747447726f77615364477277ae61747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747447726f7761747446863a5f61747447726f7761747447726f7761747447bc72"

If I change the last byte from 0x72 to 0x71, and run again the onscreen result is identical, Except the
     - Growatt plain data: last line     00000000000ce1d changed to          00000000000ce1e
 No message that the CRC is wrong.
Btw: this is the output from a MOD5000TL3-XH

**Comment by johanmeijer** [comment]
Yep your right. Grottest py is only checking the data record processing (grottdata.py)  

The crc checking is being done in grottproxy (that calls grottdata for processing the data records)  and that handles the communication.

I have an other test program that can handle the complete chain. It is really dirty coding but I will publish it when I am home again.
